### PR TITLE
i#3044 AArch64 SVE codec: Add constrained predicate count instructions

### PR DESCRIPTION
**Pull request by joshua-warburton**
This patch adds the appropriate macros, tests and codec entries to encode the following variants:
```
CNT     <Zd>.<Ts>, <Pg>/M, <Zn>.<Ts>
CNTB    <Xd>{, <pattern>{, MUL #<imm>}}
DECB    <Xdn>{, <pattern>{, MUL #<imm>}}
DECD    <Xdn>{, <pattern>{, MUL #<imm>}}
DECD    <Zdn>.D{, <pattern>{, MUL #<imm>}}
DECH    <Xdn>{, <pattern>{, MUL #<imm>}}
DECH    <Zdn>.H{, <pattern>{, MUL #<imm>}}
SQDECH  <Xdn>, <Wdn>{, <pattern>{, MUL #<imm>}}
SQDECH  <Xdn>{, <pattern>{, MUL #<imm>}}
SQDECH  <Zdn>.H{, <pattern>{, MUL #<imm>}}
```
issues: #3044

### CODE REVIEW

**Comment by joshua-warburton** [comment]
Hey @derekbruening, @abhinav92003, do you have any idea why the file_io tests are failing? Doesn't seem to be anything to do with the content of this PR as it's also failing on other PRs. Thanks

```
tart 116: code_api|client.file_io
2022-12-19T14:48:07.3715131Z 
2022-12-19T14:48:07.3720035Z 116: Test command: D:\a\dynamorio\dynamorio\build_debug-internal-64\bin64\drrun.exe "-s" "90" "-quiet" "-debug" "-use_dll" "D:/a/dynamorio/dynamorio/build_debug-internal-64/lib64/debug/dynamorio.dll" "-exit0" "-stderr_mask" "0xC" "-msgbox_mask" "0" "-dumpcore_mask" "0x7d" "-staged" "-code_api" "-c" "D:/a/dynamorio/dynamorio/build_debug-internal-64/suite/tests/bin/client.file_io.dll.dll" "D:/a/dynamorio/dynamorio/suite/tests/client-interface/file_io_data.txt" "--" "D:/a/dynamorio/dynamorio/build_debug-internal-64/suite/tests/bin/client.file_io.exe"
2022-12-19T14:48:07.3721496Z 116: Working Directory: D:/a/dynamorio/dynamorio/build_debug-internal-64/suite/tests
2022-12-19T14:48:07.3721972Z 116: Test timeout computed to be: 600
2022-12-19T14:48:07.3722327Z 115: About to create thread
2022-12-19T14:48:07.3722677Z 115: kernel_xfer_event: type 2
2022-12-19T14:48:07.3722993Z 116: 0123456789
2022-12-19T14:48:07.3723272Z 116: 01234
2022-12-19T14:48:07.3723546Z 116: abcdefg
2022-12-19T14:48:07.3723805Z 116: x
2022-12-19T14:48:07.3724075Z 116: EOF
2022-12-19T14:48:07.3724344Z 116: 
2022-12-19T14:48:07.3724644Z 115: kernel_xfer_event: type 7
2022-12-19T14:48:07.3724982Z 115: kernel_xfer_event: type 5
2022-12-19T14:48:07.3725315Z 115: kernel_xfer_event: type 6
2022-12-19T14:48:07.3725647Z 115: kernel_xfer_event: type 5
2022-12-19T14:48:07.3726400Z 116: <Application D:\a\dynamorio\dynamorio\build_debug-internal-64\suite\tests\bin\client.file_io.exe (6876).  Internal Error: DynamoRIO debug check failure: D:\a\dynamorio\dynamorio\core\win32\loader.c:771 !is_dynamo_address(dcontext->app_fls_data)
2022-12-19T14:48:07.3727351Z 116: (Error occurred @0 frags in tid 2500)
2022-12-19T14:48:07.3727716Z 116: version 9.0.19345, custom build
2022-12-19T14:48:07.3728628Z 116: -no_dynamic_options -client_lib 'D:\a\dynamorio\dynamorio\build_debug-internal-64\suite\tests\bin\client.file_io.dll.dll;0;"D:/a/dynamorio/dynamorio/suite/tests/client-interface/file_io_data.txt"' -client_lib64 'D:\a\dynamorio\dynamorio\build_debug-internal-64\suite\tests\bin\client.file_io.dll.dll;0;"D:/a/dynamorio/dyna
2022-12-19T14:48:07.3729498Z 116: 0x00000097b6b1e5e0 0x000001fd792b097a
2022-12-19T14:48:07.3729861Z 116: 0x00000097b6b1e720 0x0000000000000000
2022-12-19T14:48:07.3730370Z 116: D:\a\dynamorio\dynamorio\build_debug-internal-64\lib64\debug\dynamorio.dll=0x0000000015000000
2022-12-19T14:48:07.3731034Z 116: D:\a\dynamorio\dynamorio\build_debug-internal-64\suite\tests\bin\client.file_io.dll.dll=0x00007ff676880000
2022-12-19T14:48:07.3731592Z 116: C:\Windows/system32/KERNEL32.dll=0x000001fd79090000
2022-12-19T14:48:07.3732045Z 116: C:\Windows/system32/KERNELBASE.dll=0x000001fd79220000>
2022-12-19T14:48:07.3732621Z 116/285 Test #116: code_api|client.file_io ......................................***Failed  Required regular expression not found. Regex=[^0123456789
```

**Comment by derekbruening** [comment]
> Hey @derekbruening, @abhinav92003, do you have any idea why the file_io tests are failing? Doesn't seem to be anything to do with the content of this PR as it's also failing on other PRs. Thanks
> 
> ```
> tart 116: code_api|client.file_io
> 2022-12-19T14:48:07.3726400Z 116: <Application D:\a\dynamorio\dynamorio\build_debug-internal-64\suite\tests\bin\client.file_io.exe (6876).  Internal Error: DynamoRIO debug check failure: D:\a\dynamorio\dynamorio\core\win32\loader.c:771 !is_dynamo_address(dcontext->app_fls_data)

No, maybe a Windows update on the GA CI VM's?  But that might be expected to affect a lot of tests?  Is it only this test?

File: core/ir/aarch64/instr_create_api.h [hunk 1]
```diff
@@ -7013,4 +7016,155 @@
...
+
+/**
+ * Creates a SQDECH instruction.
+ *
+ * This macro is used to encode the forms:
+ * \verbatim
+ *    SQDECH  <Xdn>, <Wdn>{, <pattern>{, MUL #<imm>}}
+ * \endverbatim
+ * \param dc   The void * dcontext used to allocate memory for the #instr_t.
+ * \param Rdn   The first source and destination  register, X (Extended, 64 bits).
```
Comment (AssadHashmi) [inline core/ir/aarch64/instr_create_api.h:7132]: `The source and destination register in which the 32 bit result is sign extended to 64 bits.`

File: core/ir/aarch64/instr_create_api.h [hunk 2]
```diff
@@ -7013,4 +7016,155 @@
...
+
+/**
+ * Creates a DECB instruction.
+ *
+ * This macro is used to encode the forms:
+ * \verbatim
+ *    DECB    <Xdn>{, <pattern>{, MUL #<imm>}}
+ * \endverbatim
+ * \param dc   The void * dcontext used to allocate memory for the #instr_t.
+ * \param Rdn   The first source and destination  register, X (Extended, 64 bits).
```
Comment (AssadHashmi) [inline core/ir/aarch64/instr_create_api.h:7057]: `The register to be decremented,`
Similarly for `DECD` below.

File: core/ir/aarch64/instr_create_api.h [hunk 3]
```diff
@@ -7013,4 +7016,155 @@
...
+
+/**
+ * Creates a DECD instruction.
+ *
+ * This macro is used to encode the forms:
+ * \verbatim
+ *    DECD    <Zdn>.D{, <pattern>{, MUL #<imm>}}
+ * \endverbatim
+ * \param dc   The void * dcontext used to allocate memory for the #instr_t.
+ * \param Zdn   The second source and destination vector register, Z (Scalable).
```
Comment (AssadHashmi) [inline core/ir/aarch64/instr_create_api.h:7087]: `The vector register to be decremented,`

**Review by AssadHashmi — CHANGES_REQUESTED** [review]

**Comment by derekbruening** [comment]
> > Hey @derekbruening, @abhinav92003, do you have any idea why the file_io tests are failing? Doesn't seem to be anything to do with the content of this PR as it's also failing on other PRs. Thanks
> > ```
> > tart 116: code_api|client.file_io
> > 2022-12-19T14:48:07.3726400Z 116: <Application D:\a\dynamorio\dynamorio\build_debug-internal-64\suite\tests\bin\client.file_io.exe (6876).  Internal Error: DynamoRIO debug check failure: D:\a\dynamorio\dynamorio\core\win32\loader.c:771 !is_dynamo_address(dcontext->app_fls_data)
> > ```
> 
> No, maybe a Windows update on the GA CI VM's? But that might be expected to affect a lot of tests? Is it only this test?

Please file an issue on this.

**Comment by joshua-warburton** [comment]
> > > Hey @derekbruening, @abhinav92003, do you have any idea why the file_io tests are failing? Doesn't seem to be anything to do with the content of this PR as it's also failing on other PRs. Thanks
> > > ```
> > > tart 116: code_api|client.file_io
> > > 2022-12-19T14:48:07.3726400Z 116: <Application D:\a\dynamorio\dynamorio\build_debug-internal-64\suite\tests\bin\client.file_io.exe (6876).  Internal Error: DynamoRIO debug check failure: D:\a\dynamorio\dynamorio\core\win32\loader.c:771 !is_dynamo_address(dcontext->app_fls_data)
> > > ```
> > 
> > 
> > No, maybe a Windows update on the GA CI VM's? But that might be expected to affect a lot of tests? Is it only this test?
> 
> Please file an issue on this.

raised as #5802

File: core/ir/aarch64/instr_create_api.h [hunk 4]
```diff
@@ -7013,4 +7016,155 @@
...
+ * Creates a CNTB instruction.
+ *
+ * This macro is used to encode the forms:
+ * \verbatim
+ *    CNTB    <Xd>{, <pattern>{, MUL #<imm>}}
+ * \endverbatim
+ * \param dc   The void * dcontext used to allocate memory for the #instr_t.
+ * \param Rd   The destination  register, X (Extended, 64 bits).
+ * \param pattern   The immediate symbolised
+ * \param imm   The immediate shiftOp
```
[the remark this answers is not in the document]
Comment (joshua-warburton) [reply core/ir/aarch64/instr_create_api.h]: done

File: core/ir/aarch64/instr_create_api.h [hunk 5]
```diff
@@ -7013,4 +7016,155 @@
...
+ * \verbatim
+ *    DECB    <Xdn>{, <pattern>{, MUL #<imm>}}
+ * \endverbatim
+ * \param dc   The void * dcontext used to allocate memory for the #instr_t.
+ * \param Rdn   The first source and destination  register, X (Extended, 64 bits).
+ * \param pattern   The immediate symbolised
+ * \param imm   The immediate shiftOp
+ */
+#define INSTR_CREATE_decb_shift(dc, Rdn, pattern, imm) \
+    instr_create_1dst_4src(dc, OP_decb, Rdn, Rdn, pattern, OPND_CREATE_MUL(), imm)
```
[the remark this answers is not in the document]
Comment (joshua-warburton) [reply core/ir/aarch64/instr_create_api.h:7062]: done

File: core/ir/aarch64/instr_create_api.h [hunk 6]
```diff
@@ -7013,4 +7016,155 @@
...
+ * This macro is used to encode the forms:
+ * \verbatim
+ *    DECD    <Zdn>.D{, <pattern>{, MUL #<imm>}}
+ * \endverbatim
+ * \param dc   The void * dcontext used to allocate memory for the #instr_t.
+ * \param Zdn   The second source and destination vector register, Z (Scalable).
+ * \param pattern   The immediate symbolised
+ * \param imm   The immediate shiftOp
+ */
+#define INSTR_CREATE_decd_sve_shift(dc, Zdn, pattern, imm) \
```
[the remark this answers is not in the document]
Comment (joshua-warburton) [reply core/ir/aarch64/instr_create_api.h]: done

File: core/ir/aarch64/instr_create_api.h (hunk 2 above)
Comment (joshua-warburton) [reply core/ir/aarch64/instr_create_api.h]: done

File: core/ir/aarch64/instr_create_api.h (hunk 3 above)
Comment (joshua-warburton) [reply core/ir/aarch64/instr_create_api.h]: done

File: core/ir/aarch64/instr_create_api.h [hunk 7]
```diff
@@ -7013,4 +7016,155 @@
...
+ * This macro is used to encode the forms:
+ * \verbatim
+ *    CNTB    <Xd>{, <pattern>{, MUL #<imm>}}
+ * \endverbatim
+ * \param dc   The void * dcontext used to allocate memory for the #instr_t.
+ * \param Rd   The destination  register, X (Extended, 64 bits).
+ * \param pattern   The immediate symbolised
+ * \param imm   The immediate shiftOp
+ */
+#define INSTR_CREATE_cntb_shift(dc, Rd, pattern, imm) \
```
[the remark this answers is not in the document]
Comment (joshua-warburton) [reply core/ir/aarch64/instr_create_api.h]: done

File: core/ir/aarch64/instr_create_api.h (hunk 1 above)
Comment (joshua-warburton) [reply core/ir/aarch64/instr_create_api.h]: done

File: core/ir/aarch64/instr_create_api.h [hunk 8]
```diff
@@ -7013,4 +7016,155 @@
...
+ * This macro is used to encode the forms:
+ * \verbatim
+ *    DECD    <Xdn>{, <pattern>{, MUL #<imm>}}
+ * \endverbatim
+ * \param dc   The void * dcontext used to allocate memory for the #instr_t.
+ * \param Rdn   The first source and destination  register, X (Extended, 64 bits).
+ * \param pattern   The immediate symbolised
+ * \param imm   The immediate shiftOp
+ */
+#define INSTR_CREATE_decd_shift(dc, Rdn, pattern, imm) \
```
[the remark this answers is not in the document]
Comment (joshua-warburton) [reply core/ir/aarch64/instr_create_api.h]: done

File: core/ir/disassemble_shared.c [hunk 9]
```diff
@@ -586,6 +617,11 @@ internal_opnd_disassemble(char *buf, size_t bufsz, size_t *sofar INOUT,
          */
         if (TEST(DR_DISASM_ARM, DYNAMO_OPTION(disasm_mask))) {
             print_to_buffer(buf, bufsz, sofar, "%s%s%d", immed_prefix(), sign, (uint)val);
+#ifdef AARCHXX
```
Comment (AssadHashmi) [inline core/ir/disassemble_shared.c:620]: `AARCH64` ?

File: core/ir/aarch64/instr_create_api.h [hunk 10]
```diff
@@ -114,6 +114,9 @@
 /** Create an operand specifying LSL, the default shift type when there is no shift. */
 #define OPND_CREATE_LSL() opnd_add_flags(OPND_CREATE_INT(DR_SHIFT_LSL), DR_OPND_IS_SHIFT)
 
+/** Create an operand specifying MUL, a multiplier for SVE predicate constraints. */
```
Comment (AssadHashmi) [inline core/ir/aarch64/instr_create_api.h:117]: I think this should be a generic multiplier operand, not just for SVE.

**Review by AssadHashmi — CHANGES_REQUESTED** [review]

File: core/ir/aarch64/instr_create_api.h [hunk 11]
```diff
@@ -7171,4 +7174,159 @@
...
+
+/**
+ * Creates a DECD instruction.
+ *
+ * This macro is used to encode the forms:
+ * \verbatim
+ *    DECD    <Xdn>{, <pattern>{, MUL #<imm>}}
+ * \endverbatim
+ * \param dc   The void * dcontext used to allocate memory for the #instr_t.
+ * \param Rdn   The vector register to be decremented, X (Extended, 64 bits).
```
[the remark this answers is not in the document]
Comment (joshua-warburton) [reply core/ir/aarch64/instr_create_api.h]: done

File: core/ir/aarch64/instr_create_api.h [hunk 12]
```diff
@@ -7171,4 +7174,159 @@
...
+
+/**
+ * Creates a SQDECH instruction.
+ *
+ * This macro is used to encode the forms:
+ * \verbatim
+ *    SQDECH  <Zdn>.H{, <pattern>{, MUL #<imm>}}
+ * \endverbatim
+ * \param dc   The void * dcontext used to allocate memory for the #instr_t.
+ * \param Zdn   The second source and destination vector register, Z (Scalable).
```
[the remark this answers is not in the document]
Comment (joshua-warburton) [reply core/ir/aarch64/instr_create_api.h]: done

File: core/ir/instr_inline_api.h [hunk 13]
```diff
@@ -299,21 +299,23 @@ opnd_create_reg_element_vector(reg_id_t r, opnd_size_t element_size)
...
 }
 
+#    ifdef AARCH64
 INSTR_INLINE
 opnd_t
 opnd_create_predicate_reg(reg_id_t r, bool is_merge)
 {
     opnd_t opnd DR_IF_DEBUG(= { 0 }); /* FIXME: Needed until i#417 is fixed. */
-#    ifdef AARCH64
+#        ifdef AARCH64
```
[the remark this answers is not in the document]
Comment (joshua-warburton) [reply core/ir/instr_inline_api.h]: done

File: core/ir/aarch64/instr_create_api.h (hunk 10 above)
Comment (joshua-warburton) [reply core/ir/aarch64/instr_create_api.h]: done

File: core/ir/disassemble_shared.c [hunk 14]
```diff
@@ -696,7 +696,7 @@ internal_opnd_disassemble(char *buf, size_t bufsz, size_t *sofar INOUT,
         break;
     case REG_kind:
         reg_disassemble(buf, bufsz, sofar, opnd_get_reg(opnd), opnd_get_flags(opnd), "",
-                        IF_AARCHXX_ELSE(aarch64_reg_opnd_suffix(opnd), ""));
+                        IF_AARCH64_ELSE(aarch64_reg_opnd_suffix(opnd), ""));
```
Comment (AssadHashmi) [inline core/ir/disassemble_shared.c:699]: This should remain as `IF_AARCHXX_ELSE`

File: core/ir/disassemble_shared.c [hunk 15]
```diff
@@ -255,7 +255,7 @@ opnd_size_suffix_intel(opnd_t opnd)
     return "";
 }
 
-#ifdef AARCHXX
+#ifdef AARCH64
```
Comment (AssadHashmi) [inline core/ir/disassemble_shared.c:258]: This should remain as `AARCHXX`.

File: core/ir/disassemble_shared.c [hunk 16]
```diff
@@ -617,7 +617,7 @@ internal_opnd_disassemble(char *buf, size_t bufsz, size_t *sofar INOUT,
          */
         if (TEST(DR_DISASM_ARM, DYNAMO_OPTION(disasm_mask))) {
             print_to_buffer(buf, bufsz, sofar, "%s%s%d", immed_prefix(), sign, (uint)val);
-#ifdef AARCHXX
+#ifdef AARCH64
```
Comment (AssadHashmi) [inline core/ir/disassemble_shared.c:620]: This has correctly been changed to `AARCH64`.

**Review by AssadHashmi — CHANGES_REQUESTED** [review]

**Comment by joshua-warburton** [comment]
run arm tests

**Review by AssadHashmi — APPROVED** [review]

**Pull request merged**